projects/graph/graph.py

```diff
@@ -7,53 +7,100 @@ class Graph:
-        """
-        pass  # TODO
-    def dft_recursive(self, starting_vertex):
-        """
-        Print each vertex in depth-first order
-        beginning from starting_vertex.
-        This should be done using recursion.
-        """
-        pass  # TODO
-    def bfs(self, starting_vertex, destination_vertex):
-        """
-        Return a list containing the shortest path from
-        starting_vertex to destination_vertex in
-        breath-first order.
-        """
-        pass  # TODO
-    def dfs(self, starting_vertex, destination_vertex):
-        """
-        Return a list containing a path from
-        starting_vertex to destination_vertex in
-        depth-first order.
-        """
-        pass  # TODO
+        stack = Stack()
+        visited = set()
+        stack.push(starting_vertex)
+        while stack.size() > 0:
+            v = stack.pop()
+            if v not in visited:
+                visited.add(v)
+                for next_vert in self.vertices[v]:
+                    stack.push(next_vert)
+        return visited
 
+# --------------------------------------------------------------------
 
+    def dft_recursive(self, starting_vertex, visited=None):
+        if visited is None:
+            visited = set()
+        if starting_vertex not in visited:
+            print(starting_vertex)
+            visited.add(starting_vertex)
+            for child_vert in self.vertices[starting_vertex]:
+                self.dft_recursive(child_vert, visited)
+        
+        
+# --------------------------------------------------------------------
 
+    def bfs(self, starting_vertex, destination_vertex):
+        queue = Queue()
+        queue.enqueue([starting_vertex])
+        visited = set()
+        while queue.size() > 0:
+            path = queue.dequeue()
+            node = path[-1]
+            if node not in visited:
+                if node == destination_vertex:
+                    return path    
+                visited.add(node)
+                for next_node in self.vertices[node]:
+                    new_path = path.copy()
+                    new_path.append(next_node)
+                    queue.enqueue(new_path)
+        return None
 
+# --------------------------------------------------------------------
+
+    def dfs(self, starting_vertex, destination_vertex):
+        stack = Stack()
+        stack.push([starting_vertex])
+        visited = set()
+        while stack.size() > 0:
+            path = stack.pop()
+            node = path[-1]
+            if node not in visited:
+                if node == destination_vertex:
+                    return path
+                visited.add(node)
+                for next_node in self.vertices[node]:
+                    new_path = path.copy()
+                    new_path.append(next_node)
+                    stack.push(new_path)
+        return None
+
+# --------------------------------------------------------------------
```

Comment:
Nice work with the Graphs implementation.  I like your structural comments to break up the methods, it makes things easy to read.  Isnt it crazy how similar the code for BFS and DFS is?  If we really wanted in the future we could make a function that handles both DFS and BFS based on a parameter we pass in.